freqtrade/freqtradebot.py

```diff
@@ -511,38 +512,59 @@ def check_and_call_adjust_trade_position(self, trade: Trade):
+        min_exit_stake = self.exchange.get_min_pair_stake_amount(trade.pair,
+                                                                 current_exit_rate,
+                                                                 self.strategy.stoploss)
+        max_entry_stake = self.exchange.get_max_pair_stake_amount(trade.pair, current_entry_rate)
+        max_exit_stake = self.exchange.get_max_pair_stake_amount(trade.pair, current_exit_rate)
         stake_available = self.wallets.get_available_stake_amount()
         logger.debug(f"Calling adjust_trade_position for pair {trade.pair}")
         stake_amount = strategy_safe_wrapper(self.strategy.adjust_trade_position,
                                              default_retval=None)(
-            trade=trade, current_time=datetime.now(timezone.utc), current_rate=current_rate,
-            current_profit=current_profit, min_stake=min_stake_amount,
-            max_stake=min(max_stake_amount, stake_available))
+            trade=trade, current_time=datetime.now(timezone.utc), current_rate=current_entry_rate,
+            current_profit=current_entry_profit, min_stake=min_entry_stake,
+            max_stake=min(max_entry_stake, stake_available),
+            current_entry_rate=current_entry_rate, current_exit_rate=current_exit_rate,
+            current_entry_profit=current_entry_profit, current_exit_profit=current_exit_profit,
+            min_entry_stake=min_entry_stake, min_exit_stake=min_exit_stake,
+            max_entry_stake=min(max_entry_stake, stake_available),
+            max_exit_stake=min(max_exit_stake, stake_available)
+        )
 
         if stake_amount is not None and stake_amount > 0.0:
             # We should increase our position
-            self.execute_entry(trade.pair, stake_amount, trade=trade, is_short=trade.is_short)
+            if self.strategy.max_entry_position_adjustment > -1:
+                count_of_entries = trade.nr_of_successful_entries
+                if count_of_entries > self.strategy.max_entry_position_adjustment:
+                    logger.debug(f"Max adjustment entries for {trade.pair} has been reached.")
+                    return
+                else:
+                    logger.debug("Max adjustment entries is set to unlimited.")
+            self.execute_entry(trade.pair, stake_amount, None and current_entry_rate,
+                               trade=trade, is_short=trade.is_short)
 
         if stake_amount is not None and stake_amount < 0.0:
             # We should decrease our position
-            # TODO: Selling part of the trade not implemented yet.
-            logger.error(f"Unable to decrease trade position / sell partially"
-                         f" for pair {trade.pair}, feature not implemented.")
+            # Strategy should return value as Decimal for accuracy.
+            amount = abs(float(Decimal(stake_amount) / Decimal(current_exit_rate)))
```

Comment:
While the calculations themselfs are covered by tests - you're missing tests for the most critical part - which is this whole if statement, as well as the corresponding one in backtesting.

Reply:
Any test cases added for rebuys in `test_backtest_detail.py`?

Reply:
backtesting rebuy testcase is in `test_backtesting_adjust_position.py`.
i'm however not sure if that'll be easier or more difficult - as this relies on the return value from the TestStrategy - so messing with that MAY change a lot of other tests, too.

Reply:
trade details for the StrategyTestV2?

Reply:
seems so - but it's better to use StrategyTestV3  - as that's the new strategy format. I'll update the existing test to that as well.